src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -189,13 +191,23 @@ public void run() {
                         writeThread();
                     } finally {
                         state.release(StateHelper.WRITING);
-                        if (state.compareAndSet(0, StateHelper.RECONNECT))
-                            LockSupport.unpark(connector);
+                        // avoid a case when this thread falls asleep here
+                        // after WRITING flag released and then can pollute the state
+                        if (generationSync.decrementAndGet() == 0) {
+                            state.trySignalForReconnection();
+                        }
                     }
                 }
             }
         });
 
+        // reconnection preparation is done
+        // before reconnection the state will be released
+        // reader/writer threads have been replaced by new ones
+        // it's required to be sure that old r/w threads see correct
+        // client's r/w references
```

Comment:
We don't check references anymore, so the comment seems to be outdated. I would say here that at this point (just before a new reader and writer are started) we allow them to set READING / WRITING state. I would also say that it is likely (but don't guaranteed) that old threads are gone at the moment.

I tried to describe changes myself as you see above and this makes me think that bad things are still possible, because we don't ensure old threads cannot perform reading / writing of a new socket and cannot change a state somehow (say, after we release RECONNECT). However much time is spent here, most often issues are fixes and I don't want to block it anymore. I just want this possibility to be stated in a commit message and then we'll work on it within a separate issue.

Reply:
> These changes do not guarantee that a stalled reader/writer thread from a previous connection iteration will not try to read/write something from/into a new socket and will not try to change a state, but highly descrease probability of this situation.

We introduced the `leftIoThreads` variable to start reconnection after the old IO threads became harmless (at the end of their lives). Consequently, new threads cannot be started before the previous ones are still alive. Maybe I missed something out, didn't I?

Reply:
Understood now. We guarantee that a new reconnection iterator will not start until both threads will exit.